### ISSUE TITLE
incorrect check for anomalous-backslash-in-string W1401

### ISSUE THREAD

**Issue by panewt** [pylint-dev/pylint#8924]
### Bug description

Dear developers:

I am running pylint-3 on code having a line containing an re.sub containing a '\|' as the search token. Pylint-3 barks with the warning about incorrect check for anomalous-backslash-in-string W1401. However, in the documentation for re.sub, it notes that '|' a special character and says "To match a literal '|', use \|, or enclose it inside a character class, as in [|]". I checked and both '\|' and '[|]' work with python 3.9 (running on rocky 9) but the former gives the above-mentioned warning while the latter passes pylint-3.

I am using pylint-3 after whatever happened on 04aug23 per dates on documentation just got updated. Running dnf/yum doesn't say I need to update but I might be testing in the middle of a push

I did not see it in GitHub issues

Thanks,
Paul

### Command used

```shell
pylint-3
```

### Pylint output

```shell
CENSUS_household.py:2518:21: W1401: Anomalous backslash in string: '\|'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
```

### Expected behavior

Both '\|' and '[|]' per python 3.9 documentation for the special character '|' passing with no warning

### Pylint version

```shell
[paul@grotto ~]$ pylint-3 --version
pylint 2.13.7
astroid 2.11.6
Python 3.9.16 (main, May 29 2023, 00:00:00) 
[GCC 11.3.1 20221121 (Red Hat 11.3.1-4)]
[paul@grotto ~]$
```

### OS / Environment

Rocky 9.2 fully updated

**Comment by Pierre-Sassoulas** [comment]
Could you install the latest pylint with ``pip install pylint -U``. pylint 2.13.7 is old at this point, the latest is 2.17.5 or 3.0.0a6.  I don't know what pylint3 is an alias for in your environment but pylitn 3.0 is not out yet. Could you also provide an actual snippet of your code instead of a text description, please ?

**Comment by panewt** [comment]
Pierre:

I replied to your email with sample program and explanation of my environment and why my python/pylint is per whatever Rocky/RHEL/Centos. It occurred to me today that you might not be seeing the email and I need to give the info inside this GitHub site. If so, my apologies.

I am running rocky9.2 (as in bug-for-bug copy of Linux Red Hat 9.2) and their installation is 3.9 (/bin/python3 is a symlink to /bin/python3.9). Their installation of pylint is 3.9 (/bin/pylint-3 is a symlink to /bin/pylint-3.9). Obviously, pylint-3 --version info of pylint 2.13.7 bears little or no relationship to their naming convention ... which appears to be focused on version matching everything to "3.9".

Rather than trying to install a second version of material in a local directory, I just decided to write you an example program. Since the "attach file paste here" doesn't let me attach an actual working script, I am going to cut-and-paste it here along with my output from running python and pylint:

Thanks,
Paul
=============================================================
 The code:

    #!/usr/bin/python3 -B

    import re
    import sys

    # NOTE: I am on rocky9.2 and "pylint-3" is a symlink to a /usr/bin/pylint3.9
    #       which gives "pylint 2.13.7: when asked its version
    #

    def main():
        # NOTE: to keep focus, ignoring f-string warnings as they are new to me
        #       per pylint-3 warnings as of 03aug23
        #
        lineToProcess = 'We want to change the | to something else'
        sys.stdout.write ('lineToProcess = >%s<\n' % lineToProcess) # pylint: disable=C0209
        sys.stdout.write ('\n')

        # this produces garbage as the '|' is a "special character as far as
        # re is concerned
        #
        # NOTE: the "re.sub' line produces no warnings in pylint-3
        #
        processedLine_1 = re.sub ('|', '"something else"', lineToProcess)
        sys.stdout.write ('processedLine = >%s<\n' % processedLine_1) # pylint: disable=C0209
        sys.stdout.write ('\n')

        # this produces correct result
        #
        # THE BUG: using pylint-3, this line gives the following warning:
        #              test.py:31:28: W1401: Anomalous backslash in string: '\|'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
        processedLine_2 = re.sub ('\|', '"something else by escaping |"', lineToProcess)
        sys.stdout.write ('processedLine = >%s<\n' % processedLine_2) # pylint: disable=C0209
        sys.stdout.write ('\n')

        # this produces correct result
        #
        # NOTE: the "re.sub' line produces no warnings in pylint-3
        #
        processedLine_3 = re.sub ('[|]', '"something else with [|]"', lineToProcess)
        sys.stdout.write ('processedLine = >%s<\n' % processedLine_3) # pylint: disable=C0209

    # -----------------------------------------------------------------------------
    if __name__ == '__main__':
        main()

Running python:

    [paul@grotto bin]$ test.py
    lineToProcess = >We want to change the | to something else<

    processedLine = >"something else"W"something else"e"something else" "something else"w"something else"a"something else"n"something else"t"something else" "something else"t"something else"o"something else" "something else"c"something else"h"something else"a"something else"n"something else"g"something else"e"something else" "something else"t"something else"h"something else"e"something else" "something else"|"something else" "something else"t"something else"o"something else" "something else"s"something else"o"something else"m"something else"e"something else"t"something else"h"something else"i"something else"n"something else"g"something else" "something else"e"something else"l"something else"s"something else"e"something else"<

    processedLine = >We want to change the "something else by escaping |" to something else<

    processedLine = >We want to change the "something else with [|]" to something else<
    [paul@grotto bin]$

Running pylint:

    [paul@grotto bin]$ pylint-3 test.py
    ************* Module test
    test.py:31:28: W1401: Anomalous backslash in string: '\|'. String constant might be missing an r prefix. (anomalous-backslash-in-string)

    ------------------------------------------------------------------
    Your code has been rated at 9.38/10 (previous run: 9.38/10, +0.00)

    [paul@grotto bin]$

end of material

**Comment by Pierre-Sassoulas** [comment]
Yeah sorry I  don't think I received your email. Also you seem to be using a debian packaged pylint, using pip would bring the latest version. Regarding your current problem you need to add an `r` prefix to your regex string, like so : `` re.sub (r'\|', '"something else by escaping |"', lineToProcess)``, see the [documentation](https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals)

**Comment by panewt** [comment]
Pierre-Sassoulas:

Thanks for reply

Two things:

First, using pip to get the latest version of pylint will cause Rocky9 yum/dnf to report this as "out of repository" which is noise I don't want. Thanks for the suggestion, but I can't mess with the repository as I am not a sysAdmin pay grade and therefore don't understand what I am getting into when I diverge from the main path

Second, I read through this thread and cannot tell whether what I report is a real issue or just noise. You suggested "(r'\|', [...]" which I appreciate but don't understand what the 'r' means. I have a workaround using '[|]' which I understand how it is being parsed. If you can, I would like to know if I found a real issue or if I was just plain screwing up in my usage

Thanks in advance,
Paul

**Comment by Pierre-Sassoulas** [comment]
Could you check pylint's documentation and tell me if something isn't clear please (just upgraded it) ? https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/anomalous-backslash-in-string.html

**Comment by panewt** [comment]
Pierre-Sassoulas:

The page is certainly better than before. The problem still exists that the re module gives different instructions (either use '\|' without any mention of 'r' or [|]). I would think most users would expect the usage in python 3 and pylint to match. If python 3 re documentation (https://docs.python.org/3/library/re.html) says all I have to do is use '\|' then I would expect pylint to accept that. Maybe the solution is that pylint folks have to get python folk to change their documentation?

Thanks for spending the time to try to sort this out
Paul

**Comment by Pierre-Sassoulas** [comment]
The instruction is at the beginning and apply for the whole doc I think:

> The solution is to use Python’s raw string notation for regular expression patterns; backslashes are not handled in any special way in a string literal prefixed with 'r'. So r"\n" is a two-character string containing '\' and 'n', while "\n" is a one-character string containing a newline. Usually patterns will be expressed in Python code using this raw string notation.

pylint accept two possible ways to do what the documentation suggest but the r prefix is the easier one (r like raw, or regex if you need to remember it):
```python
# pylint: disable=missing-docstring
import re


def r_prefix(lineToProcess):
    re.sub (r'\|', '"something else by escaping |"', lineToProcess)

def double_escape(lineToProcess):
    re.sub ('\\|', '"something else by escaping |"', lineToProcess)
```

**Comment by panewt** [comment]
Ah, that clears things up. I disagree with "a paragraph at the top that applies to whole document" (especially not highlighted) superceding directions in the "special characters" under the exact character I am trying to work with. But that is not a pylint issues so I will not bother you with it

Thanks for pointing out that paragraph as it probably will answer future "huh?" questions on my part

Paul
